Small Improvements to RedpandaLicense feature

graphcareful opened the pull request:
## Cover letter

Included in this patchset are two quality-of-life improvements to the internal redpanda license infrastructure.
1. Make PUT_license totally idempotent
    - PUT calls of the same identical license will do nothing. Previously another controller log message was written.
2. Have GET_license return the checksum of the current loaded license

Fixes: #7119

## example via rpk
```
➜  redpanda git:(license-improvements) ✗ rpk cluster license info  --format=json
{
  "Organization": "redpanda-testing",
  "Type": "enterprise",
  "Expires": "Jul 11 2122",
  "ChecksumSHA256": "2730125070a934ca1067ed073d7159acc9975dc61015892308aae186f7455daf"
}
```

## release notes
- Includes changes to make /v1/features/license (PUT) call totally idempotent
- Includes changes to make /v/1features/license (GET) to include the checksum of the loaded license in response

twmb reviewed (approved):
+1 rpk code, some questions though

twmb commented:
Is there a reason to separate the checksum into its own http response, or could we reuse the prior response and add a field?

graphcareful commented on src/v/redpanda/admin_server.cc:
```diff
@@ -1836,6 +1836,13 @@ void admin_server::register_features_routes() {
+              if (loaded_license && (*loaded_license == license)) {
+                  /// Loaded license is idential to license in request, do
+                  /// nothing and return 200(OK)
+                  co_return ss::json::json_void();
```
I was thinking about uploading HTTP 201 to indicate that the requested license is already set, however seems a bit complicated, i'll need to use the `register_routes_raw` method and manually set the reply data and code. 

Anyone opposed to just returning 200? The small negative is the client will just observe a successful PUT even though technically nothing was done.

graphcareful commented:
> Is there a reason to separate the checksum into its own http response, or could we reuse the prior response and add a field? Is there a reason a person would want to print only the checksum, vs also always printing org, expires, type?

The only reason i thought about doing it this way is because to those who would want to know the checksum, the other information really isn't necessary. However I agree, its fine to just marshall it forward within the license struct. Doing this probably will make the change even smaller +1 on that

jcsp commented:
Please use a Fixes line in the pull request, and write the release notes in user-facing language: e.g. rather than naming our internal put_license function, give the API path

jcsp commented on src/v/redpanda/admin_server.cc:
```diff
@@ -1836,6 +1836,13 @@ void admin_server::register_features_routes() {
                   throw ss::httpd::bad_request_exception(
                     fmt::format("License is expired: {}", license));
               }
+              const auto& ft = _controller->get_feature_table().local();
```
commit message typo: s/No/Do/

jcsp commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -53,8 +53,9 @@ func newInfoCommand(fs afero.Fs) *cobra.Command {
+						Checksum     int64
 						Expired      bool `json:"license_expired,omitempty"`
-					}{info.Properties.Organization, info.Properties.Type, tm, expired}, "", "  ")
+					}{info.Properties.Organization, info.Properties.Type, tm, info.Properties.Checksum, expired}, "", "  ")
```
Is this info.Properties.Checksum going to be an error if someone uses the rpk license getter CLI on redpanda 22.2?

For the purposes of this PR, it might be simpler just not to include the checksum in the CLI output: it's not exactly meant for human eyeballs anyway.

graphcareful replied:
I just checked, won't crash:
```
{
  "Organization": "redpanda-testing",
  "Type": "enterprise",
  "Expires": "Jul 11 2122",
  "Checksum": 0
}
```

graphcareful replied:
So your saying if format == `json` include it, otherwise, hide it?

twmb commented:
imo, checksums are fairly common to see, and out of all locations, I’d
expect it most in json output. The field can have ,omitempty if we want to
elide a 0 checksum (and if we believe we will never generate a 0 checksum).
However, I’d expect to see hex or base64 rather than a number. For 22.2,
there will be no error if the field is not present.

On Mon, Nov 7, 2022 at 11:06 John Spray ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/go/rpk/pkg/cli/cmd/cluster/license/info.go
> <https://github.com/redpanda-data/redpanda/pull/7130#discussion_r1015742981>
> :
>
> >  						Expired      bool `json:"license_expired,omitempty"`
> -					}{info.Properties.Organization, info.Properties.Type, tm, expired}, "", "  ")
> +					}{info.Properties.Organization, info.Properties.Type, tm, info.Properties.Checksum, expired}, "", "  ")
>
> Is this info.Properties.Checksum going to be an error if someone uses the
> rpk license getter CLI on redpanda 22.2?
>
> For the purposes of this PR, it might be simpler just not to include the
> checksum in the CLI output: it's not exactly meant for human eyeballs
> anyway.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/7130#pullrequestreview-1170866112>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJZIJPWCLZQT45YQFSRDZ3WHFALNANCNFSM6AAAAAARZKU52M>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

emaxerrno commented:
@graphcareful  the checksum should be the name of the csum func. like so

```
{
....
"xxhash32":  3048440613,
...
}
```

emaxerrno commented:
should we backport this too?

jcsp commented:
> should we backport this too?

Maybe... @pvsune , what is the operator story for licenses and 22.2.x?  if operator just knows when it is talking to 22.2.x and ignores license files for those clusters, i think that's probably okay.

graphcareful commented:
Removed checksum output from rpk's text representation of the result, omit printing checksum in json representation if field was not reported by older version of redpanda.

Also renamed json key from `checksum` to `xxhash32`

graphcareful commented:
Modified hash to be hex representation of sha256

graphcareful commented on src/v/redpanda/admin/api-doc/features.json:
```diff
@@ -144,6 +144,10 @@
                 "expires": {
                     "type": "long",
                     "description": "Expiration date of the license in Unix epoch seconds"
+                },
```
Done

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -53,8 +53,9 @@ func newInfoCommand(fs afero.Fs) *cobra.Command {
 						Expires      string
-						Expired      bool `json:"license_expired,omitempty"`
-					}{info.Properties.Organization, info.Properties.Type, tm, expired}, "", "  ")
+						Checksum     string `json:"ChecksumSHA256,omitempty"`
```
@graphcareful pointed out the casing inconsistency to me in this response -- I think we should switch all of these to lower snake case,
```go
						Organization string `json:"organization"`
						Type         string `json:"type"`
						Expires      string `json:"expires"`
						Checksum     string `json:"checksum_sha256,omitempty"`
						Expired      bool   `json:"license_expired,omitempty"`
					}{info.Properties.Organization, info.Properties.Type, tm, info.Properties.Checksum, expired}, "", "  ")
```
